tools/serve/serve.py

```diff
@@ -1110,36 +1087,40 @@ def handle_signal(signum, frame):
 
         if route_builder is None:
             route_builder = get_route_builder
-        routes = route_builder(config.aliases, config).get_routes()
+        routes = route_builder(logger, config.aliases, config).get_routes()
 
         if config["check_subdomains"]:
-            check_subdomains(config, routes, mp_context)
+            check_subdomains(logger, config, routes, mp_context, log_handlers)
 
         stash_address = None
         if bind_address:
             stash_address = (config.server_host, get_port(""))
             logger.debug("Going to use port %d for stash" % stash_address[1])
 
         with stash.StashServer(stash_address, authkey=str(uuid.uuid4())):
-            servers = start(config, routes, mp_context, **kwargs)
-            signal.signal(signal.SIGTERM, handle_signal)
-            signal.signal(signal.SIGINT, handle_signal)
+            servers = start(logger, config, routes, mp_context, log_handlers, **kwargs)
 
-            while (all(subproc.is_alive() for subproc in iter_procs(servers)) and
-                   not received_signal.is_set() and not kwargs["exit_after_start"]):
-                for subproc in iter_procs(servers):
-                    subproc.join(1)
+            try:
+                while (all(server.proc.is_alive() for server in iter_servers(servers)) and
```

Comment:
I think the idea is that if one server fails to start, or otherwise shuts down, we want to shut down everything rather than keeping going in a partially functional state.

Reply:
I've rearranged the code in a way that's hopefully clearer.

Reply:
Aha, that's *much* clearer. I thought this was the cleanup code on shutdown, and it seemed very weird wait for servers exit if *all* of them were still running as opposed to if *any* of them were...